src/fiber/fiber.ml

```diff
@@ -1016,3 +1016,98 @@ module Expert = struct
 
   let resume a x k = resume a x k
 end
+
+module Cancellation = struct
```

Comment:
Could this at least go into a separate module in the library? Putting everything in one file doesn't seem nice.

Reply:
I don't mind, but it would actually take a bit more planning. Something obvious would look like this:

```
$ ls
  dune fiber.ml core.ml core.mli cancel.ml cancel.mli
$ cat fiber.ml
include Core
module Cancel = Cancel`
```

Which I think isn't all that great. It would work better if we start splitting other bits of the code though: `Mvar`, `Svar`, etc. Do you mind all the small modules being split into separate files?

Reply:
Why do you think this split isn't great? I like it! And splitting off smaller bits like `Svar` sounds good to me too.

Reply:
I bit the bullet and split everything. Have a look

Reply:
Great! I like the resulting set of modules but I dislike the lack of individual `mli`s. Why not give each module an `mli` with a corresponding interface?

Reply:
Do you mean to move all the interfaces from `fiber.mli` to the individual modules? Or to write an additional mli for every module? I don't like the first because it makes the signatures and the resulting docs ugly - with names like `'a Core.t` for fibers. While the latter creates a bunch of boilerplate - which I don't mind too much but I know others hate it.

Reply:
I meant to do something similar to `stdune` where we have self-contained modules like `list.{ml,mli}`.

I guess you are saying that we'll need to treat `Core` specially, because you'd like `Core.t` and associated functions to be exported as `Fiber.t`, rather than `Fiber.Core.t`, etc. I don't mind leaving it as the only module without the corresponding `mli` and put all corresponding docs into `fiber.mli` if that's sufficient to make it work. (Or maybe there is better way?)

Reply:
The problem is actually putting an mli on all the other modules (other than Core). Where it would now be necessary to write `Core.t` to refer to the fiber types. And that's also how it would look like in the odoc documentation.

Reply:
>  I don't mind leaving Core as the only module without the corresponding mli and put all corresponding docs into fiber.mli if that's sufficient to make it work. (Or maybe there is better way?)

I would say that keeping all the docs in one file is preferable when possible. It makes it easier for users to browse really.

> Or maybe there is better way?

The source of the problem is that `fiber.ml` cannot refer to an internal module if the library name is `fiber`. This is because dune assumes it will be the "lib interface module". This convention always annoyed and I would prefer something like `__main__.ml` to be taken to be the library interface. If this was possible, our `Core` could remain `Fiber` and the docs would remain relatively noise free.

Reply:
> The problem is actually putting an mli on all the other modules (other than Core). Where it would now be necessary to write `Core.t` to refer to the fiber types. And that's also how it would look like in the odoc documentation.

Oh, indeed. Say, the `Ivar` module wouldn't be able to name things `Fiber.t`.

This is very sad.

Reply:
Indeed. So I propose to leave `mvar.mli`, `pool.mli` etc to contain interfaces that are private to the library and `fiber.mli` to contain all the docs.

Reply:
OK. I'm not happy with this but I see no better way.

Reply:
Yes, it's a bit ugly but it's fairly common practice. We do it with our Dune_lang and Rpc libraries. The same is done with external libraries such as Httpaf as well. If our library grows to the point where it isn't reasonable to write all the docs in one file, we can start looking for more creative ways to split things.

Reply:
Sounds good!